### ISSUE TITLE
Action input for flow matching

### ISSUE THREAD

**Issue by windbro98** [akanazawa/fpo#2]
Hi, thanks for your great work.

I'm confused about your conditional flow matching loss in your paper. Usually, a flow matching model is used to transform a Gaussian distribution into target distribution. In your paper, you use flow matching model as a policy model, and its loss is shown in equation (7-9). However, in your training procedure, I can't find the action that is performed by human experts, which is generally the target distribution. 

I dive into your codes, and it seems that you are treating the final denoised action x1 of the policy model as the target action. This confuses me a lot, because the output action by the policy model can be quite different from the expect action for the task (as the policy model is still being trained). Can you provide me the paper that supports your operation here? Thank you.

The following code comes from gridworld/models/diffusion_policy.py, line 112-137.

```python
def compute_cfm_loss(self, state_norm: torch.Tensor,
                         x1: torch.Tensor,
                         eps: torch.Tensor,
                         t: torch.Tensor) -> torch.Tensor:
        """
        Compute conditional flow matching loss.

        Args:
            state_norm: [B, D_s] normalized input state
            x1: [B, D_a] final denoised action
            eps: [B, D_a] sampled noise
            t: [B, 1] time steps

        Returns:
            loss: [B] per-sample loss
        """
        B, D_a = eps.shape
        assert x1.shape == (B, D_a), f"x1 must be [B, D_a], got {x1.shape}"
        assert state_norm.shape[0] == B, f"state_norm must have batch size {B}, got {state_norm.shape[0]}"
        assert t.shape == (B, 1), f"t must be [B, 1], got {t.shape}"

        x_t = (1 - t) * eps + t * x1  # [B, D_a]
        inp = torch.cat([state_norm, x_t, t], dim=1)  # [B, D_s + D_a + 1]
        velocity_pred = self(inp)  # [B, D_a]

        return F.mse_loss(velocity_pred, x1 - eps, reduction='none').mean(dim=1)
```

**Comment by akanazawa** [comment]
Hey, 

FPO is a RL based algorithm, so we do not need human demonstration at all. Of course you can finetune a BC'd policy but that is something we didn't exactly explore in this paper and we're looking into now. 

In this code, the x1 is the denoised action, computed in the function above, specifically in this loop: https://github.com/akanazawa/fpo/blob/142a4a8b7a254d67a71d37e041a9685850003bce/gridworld/models/diffusion_policy.py#L90

Idea is to denoise action that leads to high reward, as enabled by the FPO objective.
Hope this helps. 

A

**Comment by windbro98** [comment]
Thank you for your answer, that really helps.
